server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -0,0 +1,372 @@
+     */
+    public static class FastFilterContext {
+        private boolean missing = false; // TODO b confirm UT that can catch this
+        private boolean hasScript = false;
+        private boolean showOtherBucket = false;
+
+        private final MappedFieldType fieldType;
+
+        private long afterKey = -1L;
+        private int size = Integer.MAX_VALUE; // only used by composite aggregation for pagination
+        private Weight[] filters = null;
+
+        private final Type type;
+
+        private RoundingValuesSource valuesSource = null;
+
+        public FastFilterContext(MappedFieldType fieldType) {
+            this.fieldType = fieldType;
+            this.type = Type.DATE_HISTO;
+        }
+
+        public FastFilterContext(CompositeValuesSourceConfig[] sourceConfigs, CompositeKey rawAfterKey, List<DocValueFormat> formats) {
+            if (sourceConfigs.length == 1 && sourceConfigs[0].valuesSource() instanceof RoundingValuesSource) {
+                this.fieldType = sourceConfigs[0].fieldType();
+                this.valuesSource = (RoundingValuesSource) sourceConfigs[0].valuesSource();
+                this.missing = sourceConfigs[0].missingBucket();
+                this.hasScript = sourceConfigs[0].hasScript();
+                if (rawAfterKey != null) {
+                    assert rawAfterKey.size() == 1 && formats.size() == 1;
+                    this.afterKey = formats.get(0).parseLong(rawAfterKey.get(0).toString(), false, () -> {
+                        throw new IllegalArgumentException("now() is not supported in [after] key");
+                    });
+                }
+            } else {
+                this.fieldType = null;
+            }
+            this.type = Type.DATE_HISTO;
+        }
+
+        public FastFilterContext(Type type) {
+            this.fieldType = null;
+            this.type = type;
+        }
+
+        public DateFieldMapper.DateFieldType getFieldType() {
+            assert fieldType instanceof DateFieldMapper.DateFieldType;
+            return (DateFieldMapper.DateFieldType) fieldType;
+        }
+
+        public RoundingValuesSource getDateHistogramSource() {
+            return valuesSource;
+        }
+
+        public void setSize(int size) {
+            this.size = size;
+        }
+
+        public void setFilters(Weight[] filters) {
+            this.filters = filters;
+        }
+
+        public void setMissing(boolean missing) {
+            this.missing = missing;
+        }
+
+        public void setHasScript(boolean hasScript) {
+            this.hasScript = hasScript;
+        }
+
+        public void setShowOtherBucket(boolean showOtherBucket) {
+            this.showOtherBucket = showOtherBucket;
+        }
+
+        public boolean isRewriteable(Object parent, int subAggLength) {
+            if (parent == null && subAggLength == 0 && !missing && !hasScript) {
+                if (type == Type.FILTERS) {
+                    return !showOtherBucket;
```

Comment:
Should we ensure the field type is numeric for FilterAggregation? Or maybe we can rely on the count method in that case to short circuit the optimization

Reply:
For filter aggregation, since filters are already constructed in Aggregator, we can just try getting the count from the filters.
For range aggregation, we do need to check the data type is numeric, if it's not, the isRewriteable method in context is meant to be the gate for applying the optimization or not.